baseplate/observers/concurrency.py

```diff
@@ -0,0 +1,66 @@
+import datetime
+import time
+
+from typing import Dict
+from typing import Optional
+
+from baseplate import _ExcInfo
+from baseplate import BaseplateObserver
+from baseplate import RequestContext
+from baseplate import ServerSpan
+from baseplate import SpanObserver
+from baseplate.lib import config
+
+
+MAX_REQUEST_AGE = datetime.timedelta(minutes=1).total_seconds()
+
+
+class ConcurrencyLimitReachedError(BaseException):
+    def __init__(self, limit: int) -> None:
+        self.limit = limit
+        super().__init__(f"Too many concurrent requests (limit was {limit})")
+
+
+class ConcurrencyObserver(BaseplateObserver):
+    @classmethod
+    def from_config(cls, app_config: config.RawConfig) -> "ConcurrencyObserver":
+        cfg = config.parse_config(
+            app_config, {"concurrency_limit": config.Optional(config.Integer)}
+        )
+        return cls(cfg.concurrency_limit)
+
+    def __init__(self, limit: Optional[int]):
+        self.live_requests: Dict[int, float] = {}
+        self.limit = limit
+
+    def cull_old_requests(self) -> None:
```

Comment:
suggestion from adam: use a queue so we can stop iterating early

Reply:
Nice!